sacrebleu/metrics/bleu.py

```diff
@@ -256,10 +255,18 @@ def corpus_score(self, sys_stream: Union[str, Iterable[str]],
         # look for already-tokenized sentences
         tokenized_count = 0
 
+        # sanity checks
+        if any(len(ref_stream) != len(sys_stream) for ref_stream in ref_streams):
+            raise EOFError("System and reference streams have different lengths!")
+        if any(line is None for line in sys_stream):
+            raise EOFError("Undefined line in system stream!")
+
         fhs = [sys_stream] + ref_streams
-        for lines in zip_longest(*fhs):
-            if None in lines:
-                raise EOFError("Source and reference streams have different lengths!")
+        for lines in zip(*fhs):
+            # remove undefined references (i.e. we have fewer references for this particular sentence)
+            lines = [x for x in lines if x is not None]
+            if len(lines) < 2:  # we need at least system + 1 defined reference
```

Comment:
What if we change this to

```python
lines = [x for x in lines if x is not None and x != ""]
```

Then the command line would work fine, unless I am missing something. You could, for example, provide 100 reference files, and if there were not 100 references for sentence 10, you could just have blank lines, which would then be skipped.

Reply:
This PR solves just the API, not the command-line interface and file-format.
My idea was (see https://github.com/mjpost/sacrebleu/issues/130#issuecomment-757051390) that
* In the case of tab-separated refs in a single file, we would just specify a range of the number of references (`--num-refs=1-5`), but no empty strings would be needed. Missing columns would be substituted with `None` automatically before calling the API.
* In the case of multiple references in multiple files, we would need empty lines for missing references, but these would be again converted to `None` before calling the API.

The advantage is that the API could be still used by someone who needs empty references for some segments. This is unusual in MT, but BLEU (and `sacrebleu`) may be used also in other tasks.

That said, I don't have strong opinion on this. If we add `and x != ""` now, the feature will be immediately available also from the command line and no further PRs will be needed (just a documentation).

Reply:
(FYI: `lines = [x for x in lines if x]` should be enough to test for both emptyness and None)

Reply:
Should we merge this as it is then? What's the final decision?

Reply:
Both merging as is (handling `None` only) and @mjpost's suggesting for handling also empty strings are OK for me.
I would slightly prefer the former as it allows empty references for some tasks (but I am not aware of any such task).

> `lines = [x for x in lines if x]` should be enough to test for both emptyness and None

Yes, but I find the explicit `if x is not None and x != ""` better because it is self-documenting that these are the two expected use cases (and `False`, empty containers and numerical zero are not expected).

Reply:
As @martinpopel suggests, I like the more explicit version.

Reply:
Just after merging this, I figured out that this also removes empty hypotheses, which could have been produced by a decoder that only generated <EOS> token in the early training phase. In the case of multi-references, first ref would then be taken as the `hyp`, leading to erroneous computation:

```python
In [22]: sys
Out[22]: 
['The dog bit the man.',
 "It wasn't surprising.",
 'The man had just bitten him.']

In [23]: refs
Out[23]: 
[['The dog bit the man.', 'It was not unexpected.', 'The man bit him first.'],
 ['The dog had bit the man.',
  'No one was surprised.',
  'The man had bitten the dog.']]

In [24]: empty_sys
Out[24]: ['', '', '']

In [25]: sacrebleu.corpus_bleu(sys, refs)
Out[25]: BLEU = 48.53 82.4/50.0/45.5/37.5 (BP = 0.943 ratio = 0.944 hyp_len = 17 ref_len = 18)

In [26]: sacrebleu.corpus_bleu(empty_sys, refs)
Out[26]: BLEU = 23.93 64.7/35.7/18.2/12.5 (BP = 0.889 ratio = 0.895 hyp_len = 17 ref_len = 19)
```

Reply:
@ozancaglayan : Good point. There's a sanity check for non-null hypotheses on line 261, but I guess that we need to allow empty hypotheses, so you can't just add `!= ''` there.

This should work:
```
lines = [lines[0]] + [x for x in lines[1:] if x is not None and x != ""]
```
But it's not pretty. Maybe it would be easier to treat the hypothesis separately from the references? I guess they're only joined together for the lowercasing and the tokenizer? 

Let me know, I can push a fix.

Reply:
Iterating over them separately would be more cleaner but probably it's done this way for tokenization and lowercasing, as they are splitted back into different iterators in line 299.

Your fix seems legitimate to me, care to add 1-2 comment lines there to explain it as well?

I hope one day, we'll refactor these I/O pre-processing stuff out of the core computation methods, which would make things much clean.

Reply:
Very good catch! Did we have a test case that failed that triggered this? If not, we really should.

Can we do something like

```python
for sys, *refs in zip(*fhs):
   lines = [sys] + [ref for ref in refs if ref is not None and ref != ""]
```

That is also a bit ugly but at least the meaning is explicit.

Reply:
OK, I've created #134 with the fix – I hope it's OK, I'm not sure how to add it here after this has been merged.